### PR TITLE
Detect clearsky modifications

### PR DESCRIPTION
**Pull request by bhellis725**
Brief description of the problem and proposed solution (if not already fully described in the issue linked to above):

The fifth condition (`c5`) of the `clearsky.detect_clearsky` function wasn't calculated as defined in the original reference (equations 12 - 14).  This pull request also addresses part of #507 where there was an implicit assumption of minutely data.  All derivatives are now calculated by explicitly dividing by the `sample_interval`.  This does not address non-uniform time intervals.

 - [ ] Closes issue #506 (addresses part of #507)
 - [ ] I am familiar with the [contributing guidelines](http://pvlib-python.readthedocs.io/en/latest/contributing.html).
 - [ ] Fully tested. Added and/or modified tests to ensure correct behavior for all reasonable inputs. Tests (usually) must pass on the TravisCI and Appveyor testing services.
 - [ ] Updates entries to `docs/sphinx/source/api.rst` for API changes.
 - [ ] Adds description and name entries in the appropriate `docs/sphinx/source/whatsnew` file for all changes.
 - [ ] Code quality and style is sufficient. Passes ``git diff upstream/master -u -- "*.py" | flake8 --diff``
 - [ ] New code is fully documented. Includes sphinx/numpydoc compliant docstrings and comments in the code where necessary.
 - [ ] Pull request is nearly complete and ready for detailed review.

### CODE REVIEW

File: pvlib/clearsky.py [hunk 1]
```diff
@@ -687,7 +687,7 @@ def detect_clearsky(measured, clearsky, times, window_length,
         raise NotImplementedError('algorithm does not yet support unequal ' \
                                   'times. consider resampling your data.')
 
-    samples_per_window = int(window_length / sample_interval)
+    samples_per_window = int(window_length / sample_interval) + 1
```
Comment (cwhanse) [inline pvlib/clearsky.py]: Revert this change please.  `samples_per_window` is counting intervals, not endpoints. Because it's an intermediate it could be renamed to be more clear.

File: pvlib/clearsky.py [hunk 2]
```diff
@@ -697,25 +697,27 @@ def detect_clearsky(measured, clearsky, times, window_length,
     # calculate measurement statistics
     meas_mean = np.mean(measured[H], axis=0)
     meas_max = np.max(measured[H], axis=0)
-    meas_slope = np.diff(measured[H], n=1, axis=0)
+    meas_ghi_diff = np.diff(measured[H], n=1, axis=0)
+    meas_slope = np.diff(measured[H], n=1, axis=0) / sample_interval
```
Comment (cwhanse) [inline pvlib/clearsky.py]: The non-uniform time steps could be handled here by
- converting the time to UNIX timestamps
- `diff`ing the time to get time intervals
- element-wise division in the calculation of `meas_slope`
- using the `diff`ed time in the calculation of `meas_line_length` , 'clear_slope`, etc.

This could be the subject of a subsequent pull request. In hindsight, I could have separated #507 into two issues (different from 1 minute, and non-uniform time steps.)

File: pvlib/clearsky.py (hunk 1 above)
Comment (bhellis725) [reply pvlib/clearsky.py]: I believe that not adding 1 gives the incorrect number of samples per window.  For example, if my data is 30-minute frequency (`sample_interval=30`) and I want 60 minute windows (`window_length=60`), the current implementation would only give 2 points per window (when the Hankel matrix is constructed in the following lines).  In this case, 2 points per window would only span a 30 minute window, not the intended 60.

Comment (cwhanse) [reply pvlib/clearsky.py]: Your example makes my point. The algorithm operates on intervals not on points in time. The value at a timestamp is considered as the value for the following (?) interval - I'd have to look carefully at the Hankel matrix and the diff to see if we adopted a left- or right- endpoint convention.

**Comment by wholmgren** [comment]
completed in #596
